### ISSUE TITLE
Adding extra cookies make users known

### ISSUE THREAD

**Issue by ozzysong** [Automattic/wp-super-cache#620]
I developed a simple plugin to cache different pages based on user prefered currency.

I tried two different versions

One with wpsc_add_cookie:

```php
function set_user_currency_cache_cookie() {
  do_action( 'wpsc_add_cookie', 'user_currency' );
}
add_cacheaction('add_cacheaction', 'set_user_currency_cache_cookie');
```

And the other:

````php
function cache_my_page_with_set_cookie($string) {
  if (isset($_COOKIE['user_currency'])) {
    $string = 'user_currency' . $_COOKIE['user_currency'];
  }
  return $string;
}
add_cacheaction('wp_cache_get_cookies_values', 'cache_my_page_with_set_cookie');
````

Both works as expected caching each currency combination in it's own file.

The problem is it makes the user "known". So checking the option "Don’t cache pages for known users" to avoid caching for logged in users also disables caching for users with the "user_currency" cookie setted.

This doesn't make sense because the user is not unique in any way. A bunch of users have the same currency settings (EUR, USD, GBP...).

Is this by design? is it a bug?

Perhaps I'm missing something :(

**Comment by stodorovic** [comment]
It's hard to decide is it only bug :confused:.

We noticed this behavior, but there are possible side effects - backward compatibility. You could look PR #616 which should fix your issue. I'm trying to create similar plugin which handles GDPR cookies and I can't continue until we merge #616 (or do something else).

Your code should work, but I'd like to give you couple advises (which don't fix your primary issue).

If you have installed WP>=4.6 then you can use [add_action](https://developer.wordpress.org/reference/functions/add_action/) in WPSC plugin. It's better to move adding cookies later (recommended action is _init_). So, you can replace:

`add_cacheaction('add_cacheaction', 'set_user_currency_cache_cookie');` with:
`add_action( 'init', 'set_user_currency_cache_cookie' );` _(It's safe for WP>=4.6)_.

For second solution you should use concatenation to add cookie because you will overwrite entire key (other cookies if they exists). Improved example for condition:
```php
if ( ! empty( $_COOKIE['user_currency'] ) && is_string( $_COOKIE['user_currency'] ) ) {
	$string .= 'user_currency=' . strtolower( $_COOKIE['user_currency'] ) . ',';
}
```
First solution is better because WPSC adds extra cookies into `RewriteCond` and it's possible to keep expert mode.

It'll be helpful if you test #616. Please send feedback.

**Comment by ozzysong** [comment]
I updated my plugin as instructed:

````php
function set_user_currency_cache_cookie() {
  do_action( 'wpsc_add_cookie', 'user_currency' );
}
add_action('init', 'set_user_currency_cache_cookie');
````

And updated WP Super Cache with commits from #616 

Now it's working as expected 😀

Checking "Don't cache pages for known users" disables caching for admins and keeps caching for users with the "user_currency" cookie.

I tested different values for the cookie and it worked every time.

It seems that we are onto something!

**Comment by iandunn** [comment]
Looks like this was fixed by https://github.com/Automattic/wp-super-cache/pull/657, which was released in `v1.6.8`. Users need to switch their config to set `$wp_cache_not_logged_in = 2;`, though.

From the UI, that's `Advanced > Miscellaneous > Cache Restrictions >  Disable caching for logged in visitors. (Recommended)`.
